File: day07/Day07P2.kt
fun hasSolution(ans: Long, soFar: Long, nums: List<Long>): Boolean {
  if (nums.isEmpty()) {
    return ans == soFar
  }
  val n = nums.first()
  val rest = nums.drop(1)
  return hasSolution(ans, soFar * n, rest) || hasSolution(ans, soFar + n, rest) || hasSolution(ans, "$soFar$n".toLong(), rest)
}

fun lineVal(line: String): Long {
    val (ansStr, numStrs) = line.split(":")
    val ans = ansStr.toLong()
    val nums = numStrs.trim().split(" ").map { it.toLong() }
    return if (hasSolution(ans, nums.first(), nums.drop(1))) {
      ans
    } else {
      0
    }
}

fun main() {
  println(generateSequence(::readLine).map(::lineVal).sum())
}

File: day07/day07p2.py
import operator
import sys
from typing import Iterable


def has_solution(ans: int, so_far: int, nums: list[str]) -> bool:
    if not nums:
        return ans == so_far
    n, *rest = nums
    return (
        has_solution(ans, so_far * int(n), rest)
        or has_solution(ans, so_far + int(n), rest)
        or has_solution(ans, int(str(so_far) + n), rest)
    )


def line_val(line: str) -> int:
    ans, first, *nums = line.split()
    return int(ans) if has_solution(int(ans), int(first), nums) else 0


def format_lines(lines: Iterable[str]) -> Iterable[str]:
    return (line.replace(':', '').rstrip() for line in lines)


print(sum(line_val(line) for line in format_lines(sys.stdin)))
